### PR TITLE
Flexible fields assignment

### PR DESCRIPTION
**Pull request by amosbird**
Field f;

before:

f = Int64(-42);

after:

f = -42;

Proof of concept
```c++
template <typename T> struct NearestFieldType;

template <> struct NearestFieldType<int32_t>   { using Type = int64_t; };
template <> struct NearestFieldType<int64_t>  { using Type = int64_t; };
template <typename T>
decltype(auto) nearestFieldType(T && x)
{
    using U = typename NearestFieldType<std::decay_t<T>>::Type;
    if constexpr (std::is_same_v<std::decay_t<T>, U>)
        return std::forward<T>(x);
    else
        return U(x);
}

template <typename T>
class TD;

template <typename T>
class F {
public:
    F(T *&& t) {
        TD<decltype(t)> _;
    }
};

int main() {
    long x = 3;
    int y = 4;
    F<decltype(nearestFieldType(x))> f;   // lvalue by ref
    F<decltype(nearestFieldType(y))> f;   // lvalue by value
    F<decltype(nearestFieldType(32))> f;  // rvalue by value
    F<decltype(nearestFieldType(32ll))> f;  // rvalue by ref
    return 0;
}
```

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
I've added the missing comment.

**Comment by alexey-milovidov** [comment]
The only non-obvious part is that when you write something like
Field f = 1;
Int64 value will be created,
but it can be transparently inserted into ColumnUInt64.

**Comment by alexey-milovidov** [comment]
Shouldn't you prove this approach by removing explicit casts across the code?

**Comment by amosbird** [comment]
> Shouldn't you prove this approach by removing explicit casts across the code?

I planed to do that, but there seems to be no simple explicit cast assignment. Here are all copy assignments reported by clang

```
    DataTypeAggregateFunction.cpp:69:     field = String();
    DataTypeArray.cpp:50:     field = Array(size);
    DataTypeEnum.cpp:115:     field = nearestFieldType(x);
    DataTypeString.cpp:38:     field = String();
    DataTypeTuple.cpp:110:     field = Tuple(TupleBackend(size));
    DataTypesDecimal.cpp:114:     field = DecimalField(T(x), scale);
    DataTypeNumberBase.cpp:163:     field = typename NearestFieldType<FieldType>::Type(x);
    DataTypeFixedString.cpp:52:     field = String();
    ColumnAggregateFunction.cpp:273:     res = String();
    ColumnAggregateFunction.cpp:483:     min = serialized;
    ColumnAggregateFunction.cpp:484:     max = serialized;
    ColumnFixedString.cpp:317:     min = String();
    ColumnFixedString.cpp:318:     max = String();
    ColumnString.cpp:336:     min = String();
    ColumnString.cpp:337:     max = String();
    ColumnNullable.cpp:75:         res = Null();
    ColumnNullable.cpp:310:         min = Null();
    ColumnNullable.cpp:311:         max = Null();
    ColumnNullable.cpp:366:     min = Null();
    ColumnNullable.cpp:367:     max = Null();
    ColumnNullable.cpp:356:         min = typename NearestFieldType<T>::Type(cur_min);
    ColumnNullable.cpp:357:         max = typename NearestFieldType<T>::Type(cur_max);
    ColumnArray.cpp:119:     res = Array(size);
    ColumnArray.cpp:336:     min = Array();
    ColumnArray.cpp:337:     max = Array();
    ColumnTuple.cpp:78:     res = Tuple(TupleBackend(tuple_size));
    ColumnTuple.cpp:322:     min = Tuple(TupleBackend(tuple_size));
    ColumnTuple.cpp:323:     max = Tuple(TupleBackend(tuple_size));
    ColumnVector.cpp:282:         min = typename NearestFieldType<T>::Type(0);
    ColumnVector.cpp:283:         max = typename NearestFieldType<T>::Type(0);
    ColumnVector.cpp:317:     min = typename NearestFieldType<T>::Type(cur_min);
    ColumnVector.cpp:318:     max = typename NearestFieldType<T>::Type(cur_max);
    ExpressionElementParsers.cpp:455:     res = float_value;
    ExpressionElementParsers.cpp:466:             res = -static_cast<Int64>(uint_value);
    ExpressionElementParsers.cpp:468:             res = uint_value;
    ExpressionElementParsers.cpp:492:     res = x;
    FunctionsTransform.h:780:                     converted_to[i] = applyVisitor(FieldVisitorConvertToNumber<Float64>(), to[i]);
    FunctionsTransform.h:786:                     const_default_value = applyVisitor(FieldVisitorConvertToNumber<Float64>(), const_default_value);
    ColumnDecimal.cpp:216:         min = typename NearestFieldType<T>::Type(0, scale);
    ColumnDecimal.cpp:217:         max = typename NearestFieldType<T>::Type(0, scale);
    ColumnDecimal.cpp:232:     min = typename NearestFieldType<T>::Type(cur_min, scale);
    ColumnDecimal.cpp:233:     max = typename NearestFieldType<T>::Type(cur_max, scale);
    SummingSortedBlockInputStream.cpp:408:         right[col_num] = (*cursor->all_columns[col_num])[cursor->pos].template get<Array>();
    SummingSortedBlockInputStream.cpp:411:         right[col_num] = (*cursor->all_columns[col_num])[cursor->pos].template get<Array>();
    SummingSortedBlockInputStream.cpp:463:         row[col_num] = Array(merged.size());
    SummingSortedBlockInputStream.cpp:465:         row[col_num] = Array(merged.size());
```
Apart from the templated functions and nested types,  `    DataTypeEnum.cpp:115:     field = nearestFieldType(x);` is the only call site that is suitable for removing.

**Comment by amosbird** [comment]
also, `Field f = 1;` won't work as this pr is only used for copy assignments, not for constructions

**Comment by amosbird** [comment]
and yes, there are tons of construction call sites suitable for simplification https://la.wentropy.com/1_lp

**Comment by amosbird** [comment]
also note:  > but it can be transparently inserted into ColumnUInt64.

It won't be able to do that, as int types don't get lifted to uint fields. If you need unsigned numbers, `u` suffix is needs

**Comment by amosbird** [comment]
> The only non-obvious part is that when you write something like
> Field f = 1;
> Int64 value will be created,
> but it can be transparently inserted into ColumnUInt64.

In fact, this makes more sense to me as it's consistent to c/c++ itself. If the conversion needs to happen, it will happen. So making the unsafe get use static_cast instead of reinterpret_cast doesn't pay extra cost if the field is only used once, which is common. 

I'll post a partial PR first. If you approve this refactor process, I'll finish the left parts, which contains hundreds of changes.

**Comment by alexey-milovidov** [comment]
> If you approve this refactor process, I'll finish the left parts, which contains hundreds of changes.

Yes, I approve.

**Comment by alexey-milovidov** [comment]
> Do you also want it also be implicitly constructable?

Why not. (Or I don't understand the motivation of these changes.)

**Comment by amosbird** [comment]
@alexey-milovidov I've rewrite almost all call sites of Field constructions, leaving some non trivial conversions intact.

reinterpret_cast generates the same instructions as static_cast for signed and unsigned conversion

https://godbolt.org/z/wurrxV

The main non-obvious part is accidentally place floats into int columns or vice versa. That needs special care.

File: dbms/src/Columns/ColumnAggregateFunction.cpp [hunk 1]
```diff
@@ -260,7 +260,7 @@ MutableColumnPtr ColumnAggregateFunction::cloneEmpty() const
 
 Field ColumnAggregateFunction::operator[](size_t n) const
 {
-    Field field = String();
+    Field field = "";
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: This is not better.

File: dbms/src/DataTypes/DataTypeAggregateFunction.cpp [hunk 2]
```diff
@@ -257,7 +257,7 @@ MutableColumnPtr DataTypeAggregateFunction::createColumn() const
 /// Create empty state
 Field DataTypeAggregateFunction::getDefault() const
 {
-    Field field = String();
+    Field field = "";
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeAggregateFunction.cpp]: .

File: dbms/src/Columns/ColumnAggregateFunction.cpp (hunk 1 above)
Comment (amosbird) [reply dbms/src/Columns/ColumnAggregateFunction.cpp]: Ok, what's the retionale behind this?

**Pull request merged**